ar-module-explainer.md

```diff
@@ -68,6 +68,39 @@ function drawScene() {
 }
 ```
 
+### Drawing UI elements
+
+The AR module supports a variety of device form factors, including "handheld" phone AR as well as "headworn" glasses-style AR. If the application wishes to show UI elements to the user, it needs to be done differently based on the form factor. This information is exposed through `XRSession`'s `interactionSpace` attribute.
+
+For handheld devices, the UI elements probably should be drawn directly to the screen in screen-space coordinates, without applying any projections. In such a case, `interactionSpace` will report `"screen-space"`.
+ 
+For headworn devices, the UI elements probably should be drawn in the world space, perhaps at some distance from the user's head so that they may easily interact with it. In such a case, `interactionSpace` will report `"world-space"`.
+
+The <a href="https://immersive-web.github.io/dom-overlays/">WebXR DOM Overlays module</a>, if implemented for the device, is able to automatically place UI for both handheld and headworn sessions. However, it is only useful if you wish to do your UI as DOM elements as opposed to 
+
+```js
+function drawScene() {
+  // draw the rest of the scene
+
+  if (xrSession.interactionSpace == "world-space") {
+    // Add UI elements to actual scene
+  } else {
+    // Draw UI elements directly to screen
+  }
+}
+
+xrSession.onselect = function(e) {
+  if (xrSession.interactionSpace == "world-space") {
+    // intersect the target ray with the UI elements in world space
+    // to determine which UI element, if any, was selected
+  } else {
+    // determine the screen-space coordinates of the select event and
```

Comment:
Annoyingly there's no easy way to do this without doing some math on the projection matrix. Perhaps select events for `"screen"` target ray modes should also come with `(x, y)` positions?

Reply:
I feel like they won't be? cc @klausw 

But if they are, then yes, that makes sense. Perhaps we should spec more about how handheld AR should work vis a vis inputs in this spec, then.

Reply:
In the current ARCore-based implementation in Chrome, there are no DOM input events while a regular immersive-ar session is active. 

Enabling DOM Overlay mode adds DOM events for the selected overlay element, including pointer events, along with the `beforexrselect` logic to make it easier for applications to deduplicate DOM events and XR input events.

FYI, in current Chrome Canary, the transient WebXR input source for screen touch already reports x/y axis values in its gamepad attribute. These are normalized -1..1 range values, not DOM coordinates.

Example:

```
> session.inputSources[0].gamepad
Gamepad
id: ""
index: -1
connected: true
timestamp: 15682728.100000007
mapping: ""
axes: (2) [-0.3270833492279053, 0.5138888359069824]
buttons: Array(3)
0: GamepadButton {pressed: false, touched: false, value: 0}
1: GamepadButton {pressed: false, touched: false, value: 0}
2: GamepadButton {pressed: false, touched: true, value: 1}
```

Reply:
Hmm. It might be important to expose the size of the screen, then? This got discussed a bit in #9: it may be useful to expose the offsets of the screen rect from the viewer space or something.

I guess one question is: given that it's handheld AR, do we expect people to draw things in viewer space with an epsilon offset from the screen, or do we expect people to draw things directly to the output buffer? In the former case we need to know the near rect of the screen to know where the "edges" are so we can draw within them, in the latter case we need to know the near rect and framebuffer size (which you have already) so that we can map the input events to screen-space coordinates. Either way you need to know the near rect.

It can potentially be exposed as an offset of the viewer space, plus a width and height. Then you need to intersect the target ray with the resultant yz plane and see where on the screen it falls.